MailCore/Utils/Notifications/NotificationActionable.swift

```diff
@@ -0,0 +1,83 @@
+/*
+ Infomaniak Mail - iOS App
+ Copyright (C) 2023 Infomaniak Network SA
+
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Foundation
+import MailResources
+import UserNotifications
+
+/// Something that can register notifications groups and actions related to them
+public protocol NotificationActionsRegistrable {
+    /// Registers the action available on an incoming mail.
+    func registerEmailActionNotificationGroup()
+}
+
+/// Notification action group identifiers
+public enum NotificationActionGroupIdentifier {
+    /// Actions related to new email notifications
+    public static let newMail = "com.infomaniak.mail.incoming"
+}
+
+public enum NewMailActionIdentifier {
+    public static let archive = "com.infomaniak.mail.incoming.action.archive"
+    public static let delete = "com.infomaniak.mail.incoming.action.delete"
+    public static let reply = "com.infomaniak.mail.incoming.action.reply"
+}
+
+public struct NotificationActionsRegistrer: NotificationActionsRegistrable {
+    public init() {
+        // META: Keep sonar could happy
+    }
+
+    // MARK: - NotificationActionsRegistrable
+
+    public func registerEmailActionNotificationGroup() {
+        let archiveIconAsset = MailResourcesAsset.archives
+        let archiveIcon = UNNotificationActionIcon(templateImageName: archiveIconAsset.name)
+        let archiveAction = UNNotificationAction(
+            identifier: NewMailActionIdentifier.archive,
+            title: MailResourcesStrings.Localizable.actionArchive,
+            options: [.authenticationRequired],
+            icon: archiveIcon
+        )
+
+        let deleteIconAsset = MailResourcesAsset.bin
+        let deleteIcon = UNNotificationActionIcon(templateImageName: deleteIconAsset.name)
+        let deleteAction = UNNotificationAction(
```

Comment:
Can we decrease the badge icon when we delete the message?

Same when we archive? As the badge refers to the number of unread messages in the inbox

Reply:
I now update the badge to reflect the unread count after an action is performed. 

It is a bit more tricky <iOS16 since I need to call `UIApplication.shared` to update the badge, that does not work in extension mode so I had to use some DI trickery to solve this.